### ISSUE TITLE
Ideas I would like to implement

### ISSUE THREAD

**Issue by base33** [KevinJump/jumps.umbraco.usync#45]
Hi Kevin,

I'd like to implement 2 changes to uSync that would make my life a LOT easier, and avoid common problems I have at work and at home.

1. Check if things exist by checking whether the alias exists instead of a matching guid.  This would fix a lot of issues for us, and makes sense because Umbraco only works with unique aliases.
2. Ability to store usync files as JSON (configurable).  The reason I would like to do this one is because I keep getting hidden hex characters that break the xml serialiser.  These characters exist in the template files, and uSync saves the XML with the characters in them.  I have a feeling they are somehow added by git, or some editor our team is using.  But JSON would give us that extra stability.

Do you mind me adding these features to uSync?

Thanks,

Craig

**Comment by KevinJump** [comment]
Hi Craig, 

I'm more than happy for you to add any features you think will make your life better.

I would check out the developments in the latest version of uSync (https://github.com/KevinJump/uSync) it's a complete re-write of uSync for umbraco 7.3 and above. I think that might answer some of the issues - specifically:

1. The current version of uSync (i.e this one v2.x) does check for aliases (i think?) so if you're getting duplicates there is a chance it's something else. - However the new one (v3) does both GUID and alias lookup, so if the guid is missing, the alias, is looked for. if you are getting duplicates created, i would be really keen to help out and see if we can get them wiped out. 

2. A lot of the xml problems are from templates, which is a pain, because the contents of the template inside the config file are often not needed or indeed ignored (if their is a file on disk, then the contents from the usync file isn't used). So in the new version (v3) the template config no longer contains the file contents only the master, and title stuff, This means you should get a lot less of that conflict.

Having said that. if you want to implement a JSON version, then it should be doable - however a lot of the internal umbraco core api stuff that usync uses produces xml, so you probably would have to go through XML at some point.

**Comment by base33** [comment]
Okay great.  I'll test out the new version with 7.3 and see if I can reproduce any of my problems.  Hopefully I can get an exact reason for why I am seeing these problems.

A lot of these issues are caused by people manually creating datatypes in production (sneaky buggers), and when I next run a deployment uSync generally complains.

I really like this package and is really valuable.  If I find anything during my testing I'll submit a pull request with fixes or any additional features I think may help people.

Thanks!

**Comment by KevinJump** [comment]
Ahh yes, i hate it when they do that. I did start a lockdown package (uCantTouchThis), but just never went back to finish it - mainly because locking down would be a real pain in emergencies :)

**Comment by base33** [comment]
Absolutely!  It sounds like a great idea, but definitely would cause panic.  Love the name too haha! :-)

I have a few ideas for dealing with conflicts like that.  I want to try them out and will share it :-)
